lms/services/canvas.py

```diff
@@ -80,30 +83,33 @@ def assert_file_in_course(self, course_id, file_id):
 
         raise CanvasFileNotFoundInCourse(file_id)
 
-    def find_matching_file_in_course(self, course_id, file_id):
+    def find_matching_file_in_course(self, course_id, file_ids):
         """
-        Return the ID of a file in course_id that matches the file with ID file_id.
+        Return the ID of a file in course_id that matches one of the files in file_ids.
 
         Search for a file that the current Canvas user can see in course_id and
-        that matches the given `file` (same filename and size) and return the
-        matching file's ID.
+        that matches one of the files in file_id's (same filename and size) and
+        return the matching file's ID.
 
         Return None if no matching file is found.
         """
-        file = self._file_service.get(file_id, type_="canvas_file")
-
-        if not file:
-            return None
-
-        file_dicts = self._api.list_files(course_id)
-
-        for file_dict in file_dicts:
-            display_name = file_dict["display_name"]
-            size = file_dict["size"]
-            id_ = str(file_dict["id"])
-
-            if display_name == file.name and size == file.size and id_ != file.lms_id:
-                return str(file_dict["id"])
+        for file_id in file_ids:
+            file = self._file_service.get(file_id, type_="canvas_file")
+
+            if not file:
+                continue
+
+            for file_dict in self._api.list_files(course_id):
```

Comment:
We're calling `self._api.list_files(course_id)` once for each `file_id` in `file_ids` even though we actually only need to call it once total. This won't result in unnecessary network requests because there's [caching on `CanvasAPIClient.list_files()`](https://github.com/hypothesis/lms/blob/b08d57b2d76d7a76e88afab560e1d848242357c5/lms/services/canvas_api/client.py#L211).

We could just put `file_dicts = self._api.list_files(course_id)` above this `for` loop, but then it would call `list_files()` once unnecessarily in the case where none of the `file_id`'s are found in the DB and that _would_ be an unnecessary network request.

It is possible of course to call `list_files()` only when necessary and then only once but doing so complicates the logic in this method a little bit, and doesn't save much due to the caching on `list_files()`. We could do it though. It shouldn't break any tests.

Reply:
Yeah I think this is reasonable. In the previous solution I built a set of keys from the size and display name which you could check easily. In this case though we've started additionally asserting that the ids don't match, so I don't think that would be as neat as it was.

You could do a comprehension to get the files first, then do the loop etc... but I think this is fine?